File: tracks/go/fundamentals/variables-and-types/memory.go
package main

import (
	"fmt"
	"unsafe"
)

func main() {
	// ========================================================================
	// 1. PRINT ADDRESSES AND SIZES
	// Use &x for address, unsafe.Sizeof(x) for size in bytes.
	// Run this and observe the output.
	// ========================================================================

	var a int = 42
	var b float64 = 3.14
	var c bool = true
	var d string = "hello"

	fmt.Println("=== Sizes and Addresses ===")
	fmt.Printf("int     — value: %v, addr: %p, size: %d bytes\n", a, &a, unsafe.Sizeof(a))
	fmt.Printf("float64 — value: %v, addr: %p, size: %d bytes\n", b, &b, unsafe.Sizeof(b))
	fmt.Printf("bool    — value: %v, addr: %p, size: %d bytes\n", c, &c, unsafe.Sizeof(c))
	fmt.Printf("string  — value: %v, addr: %p, size: %d bytes\n", d, &d, unsafe.Sizeof(d))
	// Q: Why is string 16 bytes? What are those 16 bytes made of?
	//    8 bytes for a pointer to the character data, 8 bytes for the length.
	//    The actual characters ("hello") live elsewhere — the 16 bytes is just the header.

	// ========================================================================
	// 2. ZERO VALUES IN MEMORY
	// Declare without assigning. Print addresses and values.
	// Are zero-valued variables at different addresses than initialized ones?
	// ========================================================================

	var zeroInt int
	var zeroStr string
	var zeroBool bool

	fmt.Println("\n=== Zero Values ===")
	fmt.Printf("zeroInt  — value: %v, addr: %p\n", zeroInt, &zeroInt)
	fmt.Printf("zeroStr  — value: %q, addr: %p\n", zeroStr, &zeroStr)
	fmt.Printf("zeroBool — value: %v, addr: %p\n", zeroBool, &zeroBool)

	// ========================================================================
	// 3. POINTERS AND INDIRECTION
	// ========================================================================

	x := 42
	p := &x // p holds the address of x

	fmt.Println("\n=== Pointers ===")
	fmt.Printf("x value: %d, x addr: %p\n", x, &x)
	fmt.Printf("p value: %p, p addr: %p\n", p, &p)
	fmt.Printf("*p (dereferenced): %d\n", *p)

	*p = 100
	fmt.Printf("after *p = 100, x is now: %d\n", x)
	// Q: Explain in your own words why changing *p changed x.
	//    when we did x := 42, p := &x, at this point p is the pointer to int, holding memory address of x
	//	 and we changed the value of x, through the pointer when we did *p

	// ========================================================================
	// 4. VALUE SEMANTICS
	// Assigning a struct copies it. Prove it by printing addresses.
	// ========================================================================

	type Point struct {
		X, Y int
	}

	p1 := Point{10, 20}
	p2 := p1 // copy

	fmt.Println("\n=== Value Semantics ===")
	fmt.Printf("p1 — value: %v, addr: %p\n", p1, &p1)
	fmt.Printf("p2 — value: %v, addr: %p\n", p2, &p2)

	p2.X = 999
	fmt.Printf("after p2.X = 999:\n")
	fmt.Printf("p1.X = %d (unchanged?)\n", p1.X)
	fmt.Printf("p2.X = %d\n", p2.X)
	// Q: Are p1 and p2 at the same address? Why or why not?
	//    No. p2 := p1 copied the entire struct into a new location in memory.
	//    Go's default for structs is value semantics — assignment copies all the
	//    bytes to a new address. p2 is an independent copy, so mutating p2.X
	//    doesn't affect p1.

	// ========================================================================
	// 5. SLICES: REFERENCE-ISH SEMANTICS
	// Slices share underlying arrays. This is a common source of bugs.
	// ========================================================================

	s1 := []int{1, 2, 3}
	s2 := s1 // s2 shares the same underlying array

	fmt.Println("\n=== Slice Sharing ===")
	fmt.Printf("s1: %v\n", s1)
	fmt.Printf("s2: %v\n", s2)

	s2[0] = 999
	fmt.Printf("after s2[0] = 999:\n")
	fmt.Printf("s1: %v\n", s1)
	fmt.Printf("s2: %v\n", s2)
	// Q: Why did modifying s2 also change s1?
	//    How is this different from the struct copy above?
	//    s2 := s1 copied the slice header (pointer + len + cap), but both headers
	//    still point to the same underlying array. So s2[0] = 999 writes to the
	//    shared array, and s1 sees it too.
	//    The struct had its data inside it (value type). The slice has its data
	//    elsewhere, with just a pointer to it — so copying the header shares the data.

	// ========================================================================
	// 6. ESCAPE ANALYSIS
	// Run this file with: go build -gcflags="-m" memory.go
	// Look for "escapes to heap" in the output.
	// ========================================================================

	// Q: Which variables in this file do you think escape to the heap?
	//    Run the escape analysis and check.
	//    (answer here)
}

File: tracks/rust/fundamentals/variables-and-types/memory.rs
use std::mem;

fn main() {
    // ========================================================================
    // 1. PRINT SIZES
    // Rust gives you std::mem::size_of and size_of_val.
    // ========================================================================

    let a: i32 = 42;
    let b: f64 = 3.14;
    let c: bool = true;
    let d: char = 'A';
    let e: &str = "hello";
    let f: String = String::from("hello");

    println!("=== Sizes ===");
    println!("i32    — size: {} bytes", mem::size_of_val(&a));
    println!("f64    — size: {} bytes", mem::size_of_val(&b));
    println!("bool   — size: {} bytes", mem::size_of_val(&c));
    println!("char   — size: {} bytes", mem::size_of_val(&d));
    println!("&str   — size: {} bytes", mem::size_of_val(&e));
    println!("String — size: {} bytes", mem::size_of_val(&f));
    // Q: Why is &str 16 bytes and String 24 bytes?
    //    What are the extra 8 bytes in String?
    //    (answer here)

    // ========================================================================
    // 2. ADDRESSES
    // Use {:p} format to print pointer addresses.
    // ========================================================================

    let x: i32 = 42;
    let y: i32 = 43;

    println!("\n=== Addresses ===");
    println!("x addr: {:p}", &x);
    println!("y addr: {:p}", &y);
    // Q: How far apart are x and y in memory? Does that match their sizes?
    //    (answer here)

    // ========================================================================
    // 3. MOVE SEMANTICS
    // When you assign a heap type, ownership MOVES.
    // ========================================================================

    let s1 = String::from("hello");
    println!("\n=== Move Semantics ===");
    println!("s1 addr: {:p}, data addr: {:p}", &s1, s1.as_ptr());

    let s2 = s1; // s1 is MOVED to s2
    // println!("s1: {}", s1); // uncomment to see the compile error
    println!("s2 addr: {:p}, data addr: {:p}", &s2, s2.as_ptr());
    // Q: s2 is at a different stack address than s1 was, but the data pointer
    //    is the same. What does that tell you about what "move" actually does?
    //    (answer here)

    // ========================================================================
    // 4. COPY SEMANTICS
    // Stack types that implement Copy are bitwise copied.
    // ========================================================================

    let a: i32 = 42;
    let b = a; // copy, not move — both are still valid

    println!("\n=== Copy Semantics ===");
    println!("a: {}, addr: {:p}", a, &a);
    println!("b: {}, addr: {:p}", b, &b);
    // Q: a and b have different addresses. The value was copied.
    //    Why doesn't String implement Copy?
    //    (answer here)

    // ========================================================================
    // 5. REFERENCES (BORROWING)
    // A reference is a pointer that the borrow checker tracks.
    // ========================================================================

    let s = String::from("hello");
    let r: &String = &s;

    println!("\n=== References ===");
    println!("s addr:  {:p}", &s);
    println!("r value: {:p} (points to s)", r);
    println!("r addr:  {:p} (where r itself lives)", &r);
    println!("r size:  {} bytes (it's a pointer)", mem::size_of_val(&r));
    // Q: r is 8 bytes (a pointer). It points to s on the stack.
    //    s is 24 bytes on the stack, pointing to heap data.
    //    Draw the full picture: stack → stack → heap.
    //    (answer here)

    // ========================================================================
    // 6. SHADOWING vs MUTATION IN MEMORY
    // ========================================================================

    let x = 5;
    println!("\n=== Shadowing ===");
    println!("x = {}, addr: {:p}", x, &x);

    let x = x + 1; // shadows — new variable
    println!("x = {}, addr: {:p}", x, &x);
    // Q: Did the address change? What does that tell you about shadowing?
    //    (answer here)

    let mut y = 5;
    println!("\ny = {}, addr: {:p}", y, &y);

    y = y + 1; // mutation — same variable
    println!("y = {}, addr: {:p}", y, &y);
    // Q: Did the address change this time? Why or why not?
    //    (answer here)
}
